### PR TITLE
filterprocessor: fix filtering non-string body by bodies property

### PR DESCRIPTION
**Pull request by sumo-drosiek**
**Description:**

Convert body to string before filtering

**Link to tracking Issue:** #22736 

**Testing:** Unit tests

**Documentation:** N/A

### CODE REVIEW

**Review by TylerHelmuth — CHANGES_REQUESTED** [review]
Thank you for bringing this to our attention.

With https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18642 in the works I really really don't want to make changes to `filterspan`, `filtermetric`, or `filterlog` anymore. The ability to index `body` via OTTL has been introduced via https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/22102 and will be available in the next release of the collector to solve the problem this PR solves.

```yaml
filter:
  logs:
    log_record:
    - 'IsMatch(body["some key"], "some value")'
```

I am willing to hear arguments as to why we should also patch this in filterlog, but blocking the PR for now.

**Comment by sumo-drosiek** [comment]
> I am willing to hear arguments as to why we should also patch this in filterlog, but blocking the PR for now.

I wouldn't say this is the same. First of all, the current behavior of `bodies` is just wrong. We are rejecting all logs which aren't string. Also, the usecase is to have one common configuration for logs, where body may be both maps and strings (not sure if OTTL supports that).

> With https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18642 in the works I really really don't want to make changes to filterspan, filtermetric, or filterlog anymore. The ability to index body via OTTL has been introduced via https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/22102 and will be available in the next release of the collector to solve the problem this PR solves.

Totally understand, but raised it because IMHO it's bug in `bodies` behavior and should be either fixed or removed at all

**Comment by TylerHelmuth** [comment]
> Also, the usecase is to have one common configuration for logs, where body may be both maps and strings (not sure if OTTL supports that

OTTL currently supports fetching the log as whatever type it is and with version 0.79.0 you'll be able to index it if it is a map or a slice.  At the moment filterprocessor can take advantage of the OTTL filter interface, and soon all the other components using `internal/filter` will as well via https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18642.

> IMHO it's bug in bodies behavior

Agreed, but I feel the fix to the bug should be to get people off of this configuration and on to OTTL.  

That said, only filterprocessor has the option to use the OTTL configuration, so it isn't a viable fix for attributesprocessor (the only other component using `filterlog`), so that is a reason to move forward with this PR.

File: .chloggen/drosiek-fix-bodies-filtering.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Use this changelog template to create an entry for release notes.
+# If your change doesn't affect end users, such as a test fix or a tooling change,
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: filterprocessor
```
Comment (TylerHelmuth) [inline .chloggen/drosiek-fix-bodies-filtering.yaml:9]:
```suggestion
component: internal/filter/filterlog
```

File: .chloggen/drosiek-fix-bodies-filtering.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: fix filtering non-string body by bodies property
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [22736]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
```
Comment (TylerHelmuth) [inline .chloggen/drosiek-fix-bodies-filtering.yaml:20]:
```suggestion
subtext: Affects `filterprocessor` and `attributesprocessor`.
```

**Review by TylerHelmuth — APPROVED** [review]

**Pull request merged**